### PR TITLE
Impliment i3-style marks

### LINKED ISSUE #1007: i3 marks
http://i3wm.org/docs/userguide.html#vim_like_marks

Involves implementing the mark commands and the rendering on window borders.

### PR DESCRIPTION
**Pull request by 4e554c4c**
This commit adds three commands to sway: `show_marks`, `mark` and
`unmark`. Marks are displayed right-aligned in the window border as i3
does. Marks may be found using criteria.

Fixes #1007 
I may want to add some commits that improve criteria handling before this gets merged.

### CODE REVIEW

File: sway/commands/mark.c [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+							view->marks = NULL;
+						}
+					}
+					free(mark);
+				}
+				else {
+					list_add(view->marks, mark);
+				}
+			}
+			else {
```
Comment (SirCmpwn) [inline sway/commands/mark.c]: Join your elses on the same line as your closing braces

File: sway/commands/unmark.c [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+	if (view->marks) {
+		if (argc) {
+			char *mark = join_args(argv, argc);
+			int index;
+			if (-1 != (index = list_seq_find(view->marks, (int (*)(const void *, const void *))strcmp, mark))) {
+				free(view->marks->items[index]);
+				list_del(view->marks, index);
+
+				if (0 == view->marks->length) {
```
Comment (SirCmpwn) [inline sway/commands/unmark.c]: No yoda conditionals please

**Comment by SirCmpwn** [comment]
Looks good at first glance. Will review more in depth tomorrow. Nice work, thanks!

File: sway/commands/unmark.c (hunk 2 above)
Comment (4e554c4c) [inline sway/commands/unmark.c]: What are Yoda conditionals?

Comment (Timidger) [inline sway/commands/unmark.c]: Constant value before the variable in the comparison.
 E.g: "0 equals mark's length" instead of "mark's length equals 0"

**Comment by SirCmpwn** [comment]
Found an issue:

1. `show_marks on`
2. `mark a`
3. `show_marks off`
4. `show_marks on`

The marks do not re-appear.

File: sway/border.c [hunk 3]
```diff
@@ -190,6 +193,39 @@ static void render_title_bar(swayc_t *view, cairo_t *cr, struct wlc_geometry *b,
...
+		char *cursor = marks;
+		for (int i = 0; i < view->marks->length; ++i) {
+			char *mark = (char *)view->marks->items[i];
+			if (*mark != '_') {
+				int len = strlen(mark);
+				*cursor = '[';
+				strcpy(cursor + 1, mark);
+				*(cursor + len + 1) = ']';
+				cursor += len + 2;
+				*cursor = '\0';
```
Comment (SirCmpwn) [inline sway/border.c]: I would prefer snprintf to all of this manual string slinging

File: sway/border.c [hunk 4]
```diff
@@ -190,6 +193,39 @@ static void render_title_bar(swayc_t *view, cairo_t *cr, struct wlc_geometry *b,
...
+	if (config->show_marks && view->marks) {
+		int size = 0;
+		char *marks;
+
+		for (int i = 0; i < view->marks->length; ++i) {
+			if (*(char *)view->marks->items[i] != '_') {
+				size += strlen((char *)view->marks->items[i]) + 2;
+			}
+		}
+		marks = (size ? (char *)malloc(size + 1) : strdup("\0"));
```
Comment (SirCmpwn) [inline sway/border.c]: You can also use snprintf to measure the final string here

File: sway/border.c [hunk 5]
```diff
@@ -190,6 +193,39 @@ static void render_title_bar(swayc_t *view, cairo_t *cr, struct wlc_geometry *b,
...
+				cursor += len + 2;
+				*cursor = '\0';
+			}
+		}
+
+		int width, height;
+		get_text_size(cr, config->font, &width, &height, 1, false, "%s", marks);
+		cairo_move_to(cr, MAX((int)tb->size.w + x - (width + 2), x + 2), y + 2);
+		cairo_set_source_u32(cr, colors->text);
+		pango_printf(cr, config->font, 1, false, "%s", marks);
```
Comment (SirCmpwn) [inline sway/border.c]: Better yet, how about just doing this in a loop and avoiding allocation and string slinging alltogether?

File: sway/commands/mark.c [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+struct cmd_results *cmd_mark(int argc, char **argv) {
+	struct cmd_results *error = NULL;
+	if (config->reading) return cmd_results_new(CMD_FAILURE, "mark", "Can't be used in config file.");
+	if ((error = checkarg(argc, "floating", EXPECTED_AT_LEAST, 1))) {
+		return error;
+	}
+
+	swayc_t *view = get_focused_container(&root_container);
+	int add = false;
+	int toggle = false;
```
Comment (SirCmpwn) [inline sway/commands/mark.c]: `bool add = false`

include stdbool.h too

**Comment by 4e554c4c** [comment]
Looks like the `show_marks` issue was because of some dumb code copying. These problems should be fixed now.

**Comment by SirCmpwn** [comment]
Thanks!

**Comment by SirCmpwn** [comment]
Right, so this change is eligible for a bounty. Please follow up via sir@cmpwn.com to claim it, or you can redonate it to another bounty if you wish.

**Comment by 4e554c4c** [comment]
@SirCmpwn Please redonate my bounty to #892. Thanks for the review!

**Pull request merged**